deep4cast/forecasting.py

```diff
@@ -0,0 +1,224 @@
+# -*- coding: utf-8 -*-
+"""Time series regression module.
+
+This module provides access to forecasters that can be fit to univariate
+or multivariate time series.
+
+"""
+import numpy as np
+
+from abc import ABC, abstractmethod
+from hyperopt import hp, fmin, tpe
+from keras.optimizers import RMSprop, SGD
+from deep4cast.models import TemporalCNN, TruncatedRNN
+
+
+class Forecaster(ABC):
+    """Abstract data handler class."""
+
+    def __init__(self):
+        """Initialize properties."""
+        # Model-dependent internal variables
+        self._model = None
+        self._history = None
+
+        # Internal boolean checks
+        self.is_fitted = False
+
+    def fit(self, X, y, validation_split=0.0, verbose=0):
+        """Fit model to data.
+
+        :param X: Independent variable array.
+        :type X: numpy.array with shape (n,m,p)
+        :param y: Dependent variable array.
+        :type y: numpy.array with shape (q,r)
+        :param validation_split: Fraction of data for validation.
+        :type validation_split: float
+        :param  verbose: Toggle print to screen during fitting.
+        :type verbose: int
+
+        """
+        self.check_data_format(X, y)
+
+        # Set up the model based on internal model class which needs to be
+        # provided in the concrete implementation.
+        self._model = self._model_class(
+            input_shape=(X.shape[1], X.shape[2]),
+            topology=self._topology
+        )
+
+        # Compose and compile model with loss and optimizer, both of which
+        # need to be provided in the concerte implementation.
+        self._model.compile(
+            loss=self._loss,
+            optimizer=self._optimizer,
+            metrics=self._metrics
+        )
+
+        # Fit model to data  and set internal history object for
+        # ecapsulated access from outside.
+        self._history = self._model.fit(
+            X,
+            y,
+            batch_size=self._batch_size,
+            epochs=self._epochs,
+            verbose=verbose,
+            validation_split=validation_split
+        )
+
+        self.is_fitted = True
+
+    @abstractmethod
+    def hyperfit(self, X, y, validation_split=0):
```

Comment:
`hyperfit` should be outside the scope of this class. Take a look at this example: https://github.com/zer0n/MLx/blob/master/pymlx/parameter_sweep.py

Reply:
Will do